aten/src/ATen/native/mobile/cpu/internal/Add.cpp

```diff
@@ -0,0 +1,244 @@
+#include <ATen/native/mobile/cpu/Engine.h>
+
+#ifdef USE_XNNPACK
+
+#include <ATen/native/mobile/cpu/internal/Allocator.h>
+#include <ATen/native/mobile/cpu/internal/Common.h>
+
+namespace at {
+namespace native {
+namespace mobile {
+namespace cpu {
+namespace internal {
+namespace add {
+
+struct Context final {
+  Operator add_op;
+
+  static constexpr float kMin = -std::numeric_limits<float>::infinity();
+  static constexpr float kMax = std::numeric_limits<float>::infinity();
+};
+
+namespace {
+
+// Supports NHWC and NCHW FP32 adds.
+
+bool available(
+    const int64_t input1_channels,
+    const int64_t input2_channels,
+    const int64_t output_channels,
+    const Scalar output_min,
+    const Scalar output_max) {
+         // Mobile
+  return mobile::cpu::available() &&
+         // Input1
+         (input1_channels > 0) &&
+         // Input2
+         (input2_channels > 0) &&
+         // Input1 == Input2
+         (input1_channels == input2_channels) &&
+         // Output
+         (output_channels > 0) &&
+         // Input == Output
+         (input1_channels == output_channels) &&
+         // Output Min / Max
+         (output_min.isIntegral(true) || output_min.isFloatingPoint()) &&
+         (output_max.isIntegral(true) || output_max.isFloatingPoint()) &&
+         (output_max.to<float>() > output_min.to<float>());
```

Comment:
Educate me: what's going on with `output_min` and `output_max`? Is there some quantization going on here?

Reply:
Allowing all operators to be fused with a clamp (ReLU, ReLU6, HardTanh, or any operator can be represented as a clamp.)  It is much cheaper, both in terms of time and power, to perform the operation on the fly while the data is in registers, as opposed to writing it to memory, only to read it back to perform a simple operation.